File: C++/BinarySearch.cpp
//
//  main.cpp
//  BinarySearch
//
//  Created by Mai Tuyet Luu on 5/1/20.
//  Copyright © 2020 Mai Tuyet Luu. All rights reserved.
//
#include <algorithm>
#include <cstdlib>
#include <iostream>
using namespace std;


int BinarySearch(int arr[], int left, int right, int x);

void show(int a[], int size);

int main()
{
    int n = 100;
    int a[n];
    
    //int a[] = {1, 5, 8, 9, 6, 7, 3, 4, 2, 0};
    //int size = sizeof(a)/sizeof(a[0]);
    
    srand((unsigned)time(0));
    
    cout << "Generating an array... " << endl;
    
    for(int i = 0; i < n; i++)
    {
        a[i] = (rand() % 100) + 1;
        //cout << arr1[i] << " ";
    }
    
    cout << "\nThe array is: ";
    show(a, n);
    
    sort(a, a+n);
    
    cout << "\nThe array after sorting: ";
    show(a, n);
    
    auto start = std::chrono::high_resolution_clock::now();
    
    BinarySearch(a, 0, n-1, 2);
    
    auto end = std::chrono::high_resolution_clock::now();
       
    auto t = std::chrono::duration_cast<std::chrono::nanoseconds> (end - start).count();
    
    cout << "Time for Binary search: " << t << endl;
    
    cout << endl;
    
    
}

void show(int a[], int n)
{
    for(int i = 0; i < n; i++)
    {
        cout << a[i] << " ";
    }
}

int BinarySearch(int a[], int left, int right, int x)
{
    if (right >= left)
    {
        int mid = (left + right)/2;
        
        if(a[mid] == x)
        {
            cout << "\nThe value is found." << endl;
            return mid;
        }
        
        if(a[mid] > x)
        {
            return BinarySearch(a, left, mid - 1, x);
        }
        
        return BinarySearch(a, mid+1, right, x);
        
    }
    cout << "\nThe value is not found!" << endl;
    return -1;
}

File: python/binarySearch.py
import time


def binary_search(arr, start, end, val):
    while start <= end:
        mid = (start + end)//2
        if arr[mid] == val:
            return mid
        # val might be in right side
        elif arr[mid] < val:
            start = mid + 1
        # val might be on left
        else:
            end = mid - 1
    # value not found
    return -1


def one_dimension_test(size):
    first = []
    # create & fill array of specified size
    for i in range(size):
        first.append(i+1)
    # always look for val 500
    start = time.time_ns()
    binary_search(first, 0, len(first)-1, 500)
    end = time.time_ns()
    total = end - start
    return total


def average_one_dimension():
    test_values = (30000000, 15000000, 7500000, 3750000, 1875000)
    for val in test_values:
        sum_of_run = 0
        for i in range(10):
            sum_of_run += one_dimension_test(val)
            if i == 9:
                print("Average of 10 runs for one-dimensional array of size", val, ":", sum_of_run/10, "nanoseconds")


average_one_dimension()
